From: Suraj Jitindar Singh <surajjs@amazon.com>
Date: Wed, 17 Jan 2024 00:15:52 +0000
Subject: Revert "netfilter: ipset: fix race condition between swap/destroy and
 kernel side add/del/test"

This reverts commit 875ee3a09e27b7adb7006ca6d16faf7f33415aa5.

This commit has been reported upstream to cause a performance regression
in ip_set_swap() [1].

A fix was proposed but hasn't been sent through official channels for
inclusion in the kernel yet [2].

As such revert it for the amazon-6.1.y/mainline kernel for now to avoid
the regression.

[1] https://lore.kernel.org/all/C0829B10-EAA6-4809-874E-E1E9C05A8D84@automattic.com/
[2] https://lore.kernel.org/all/d5c24887-b2d4-bcc-f5a4-bd3d2670d16@blackhole.kfki.hu/

Signed-off-by: Suraj Jitindar Singh <surajjs@amazon.com>
---
 net/netfilter/ipset/ip_set_core.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/net/netfilter/ipset/ip_set_core.c b/net/netfilter/ipset/ip_set_core.c
--- a/net/netfilter/ipset/ip_set_core.c
+++ b/net/netfilter/ipset/ip_set_core.c
@@ -61,8 +61,6 @@ MODULE_ALIAS_NFNL_SUBSYS(NFNL_SUBSYS_IPSET);
 	ip_set_dereference((inst)->ip_set_list)[id]
 #define ip_set_ref_netlink(inst,id)	\
 	rcu_dereference_raw((inst)->ip_set_list)[id]
-#define ip_set_dereference_nfnl(p)	\
-	rcu_dereference_check(p, lockdep_nfnl_is_held(NFNL_SUBSYS_IPSET))
 
 /* The set types are implemented in modules and registered set types
  * can be found in ip_set_type_list. Adding/deleting types is
@@ -710,10 +708,15 @@ __ip_set_put_netlink(struct ip_set *set)
 static struct ip_set *
 ip_set_rcu_get(struct net *net, ip_set_id_t index)
 {
+	struct ip_set *set;
 	struct ip_set_net *inst = ip_set_pernet(net);
 
-	/* ip_set_list and the set pointer need to be protected */
-	return ip_set_dereference_nfnl(inst->ip_set_list)[index];
+	rcu_read_lock();
+	/* ip_set_list itself needs to be protected */
+	set = rcu_dereference(inst->ip_set_list)[index];
+	rcu_read_unlock();
+
+	return set;
 }
 
 static inline void
@@ -1396,9 +1399,6 @@ static int ip_set_swap(struct sk_buff *skb, const struct nfnl_info *info,
 	ip_set(inst, to_id) = from;
 	write_unlock_bh(&ip_set_ref_lock);
 
-	/* Make sure all readers of the old set pointers are completed. */
-	synchronize_rcu();
-
 	return 0;
 }
 

